src/transaction/log_recovery_redo_parallel.hpp

```diff
@@ -311,9 +156,30 @@ namespace cublog
       cubthread::entry_workpool *m_worker_pool;
       std::vector<std::unique_ptr<redo_task>> m_redo_tasks;
 
-      redo_job_queue m_job_queue;
-
       bool m_waited_for_termination;
+
+      std::atomic_bool m_adding_finished;
+
+      std::hash<VPID> m_vpid_hash;
+
+      /* the not-applied set from the outside by the main thread (for recovery, the main thread
+       * is the thread running the analysis redo and undo; for replication, the main thread is the
+       * recovery applying thread);
+       * the reason the outer log_lsa is needed is because some of the log entries are being applied
+       * synchronously by the main thread and the calculation of the minimum not-applied log_lsa needs
+       * to be accurate
+       */
+      std::atomic<log_lsa> m_outer_not_applied_log_lsa;
+
+      /* following members control the pro-active calculation of the minimum not-applied log_lsa as well
+       * as means to actually trigger and wait for the calculation asynchronously
+       */
+      log_lsa m_calculated_minimum_not_applied_log_lsa;
```

Comment:
To reduce the name size, `minimum_not_applied_log_lsa` may be replaced by `min_unapplied_lsa`.